torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1152,9 +1158,48 @@ ProcessGroupNCCL::~ProcessGroupNCCL() {
   abort(abortReason);
 }
 
+void dump_debugging_info() {
+  LOG(ERROR) << "Dumped debugging information: " << dump_nccl_trace();
+}
+
+void ProcessGroupNCCL::ncclCommsMonitor() {
+  int hearBeatCounter = -1;
+  while (true) {
+    auto heartbeat = heartbeat_.load();
+    if (heartbeat > hearBeatCounter) {
+      hearBeatCounter = heartbeat;
+      // Sleep for enough time so that heartBeat will at least increase.
+      std::this_thread::sleep_for(
+          std::chrono::milliseconds(kWatchdogThreadSleepMillis * 3));
+    } else {
+      // Timeout case and we will first dump the flight recorder
+      // to std::out, and we use a different detached thread to ensure
+      // monitor thread will never be blocked and to dump debugging info
+      // is best effort.
+      std::thread sideThread(dump_debugging_info);
+      sideThread.detach();
+      break;
+    }
+  }
+
+  // Create a error message reported from MonitorThread, so
+  // we throw exception and make the whole process to be killed.
+  const auto exitMsg = c10::str(
+      "[Rank ", rank_, "] NCCL monitor thread terminated with timeout: ");
+  monitorException_ = std::make_exception_ptr(std::runtime_error(exitMsg));
+  if (!terminateProcessGroup_.load()) {
+    LOG(ERROR) << "monitoring thread detects no heartbeat and throw a exception!";
+    std::rethrow_exception(monitorException_);
+  }
+}
+
 void ProcessGroupNCCL::ncclCommWatchdog() {
   try {
     VLOG(2) << "[Rank " << rank_ << "] NCCL watchdog thread started!";
+    if(monitorThreadEnabled_.load()) {
```

Comment:
one reason for launching the thread here instead of inside constructor is because it will let the function called by the thread can be overridden.

Reply:
So to test it in unit test, we want to capture the errors from this thread right? If we create the thread in the constructor of `ProcessGroupNCCL`. Then even if we make `ncclCommsMonitor` virtual and override it in the child class (for testing purpose), it won't work. The thread will still execute `ProcessGroupNCCL::ncclCommsMonitor`, aka the method from parent. That's why I move the thread creation here so that we can do some tricks in test/cpp/c10d/ProcessGroupNCCLErrorsTest.cpp.